core/src/poh_recorder.rs

```diff
@@ -263,35 +265,76 @@ impl PohRecorder {
-        if e.is_err() {
-            info!("WorkingBank::sender disconnected {:?}", e);
-            //revert the cache, but clear the working bank
+        if send_result.is_err() {
+            info!("WorkingBank::sender disconnected {:?}", send_result);
+            // revert the cache, but clear the working bank
             self.clear_bank();
         } else {
-            //commit the flush
-            let _ = self.tick_cache.drain(..cnt);
+            // commit the flush
+            let _ = self.tick_cache.drain(..entry_count);
         }
 
         Ok(())
     }
 
     pub fn tick(&mut self) {
-        let tick = self.generate_tick();
-        trace!("tick {}", tick.1);
+        let poh_entry = self.poh.lock().unwrap().tick();
+        if let Some(poh_entry) = poh_entry {
+            self.tick_height += 1;
+            trace!("tick {}", self.tick_height);
 
-        if self.start_leader_at_tick.is_none() {
-            return;
-        }
+            if self.start_leader_at_tick.is_none() {
+                return;
+            }
 
-        self.tick_cache.push(tick);
-        let _ = self.flush_cache(true);
+            let entry = Entry {
+                num_hashes: poh_entry.num_hashes,
+                hash: poh_entry.hash,
+                transactions: vec![],
+            };
+
+            self.tick_cache.push((entry, self.tick_height));
+            let _ = self.flush_cache(true);
+        }
     }
 
-    pub fn record(&mut self, bank_slot: u64, mixin: Hash, txs: Vec<Transaction>) -> Result<()> {
-        self.flush_cache(false)?;
-        self.record_and_send_txs(bank_slot, mixin, txs)
+    pub fn record(
+        &mut self,
+        bank_slot: u64,
+        mixin: Hash,
+        transactions: Vec<Transaction>,
+    ) -> Result<()> {
+        // Entries without transactions are used to track real-time passing in the ledger and
+        // cannot be generated by `record()`
+        assert!(!transactions.is_empty(), "No transactions provided");
+        loop {
+            self.flush_cache(false)?;
+
+            let working_bank = self
+                .working_bank
+                .as_ref()
+                .ok_or(Error::PohRecorderError(PohRecorderError::MaxHeightReached))?;
+            if bank_slot != working_bank.bank.slot() {
+                return Err(Error::PohRecorderError(PohRecorderError::MaxHeightReached));
+            }
+
+            if let Some(poh_entry) = self.poh.lock().unwrap().record(mixin) {
+                let entry = Entry {
+                    num_hashes: poh_entry.num_hashes,
+                    hash: poh_entry.hash,
+                    transactions,
+                };
+                self.sender
+                    .send((working_bank.bank.clone(), vec![(entry, self.tick_height)]))?;
+                return Ok(());
+            }
+            // record() might fail if the next PoH hash needs to be a tick.  But that's ok, tick()
+            // and re-record()
+            self.tick();
```

Comment:
Can this race with the other tick() in poh_service() and other banking threads calling record()? Multiple threads can observe 1 hash remaining, and decide to call `tick()` (which I think decrements the `remaining_hashes` counter, could potentially cause a subtraction overflow?)

Reply:
No, I don't think so.   When `remaining_hashes` hits 0 within `tick()`, it gets reset back to `self.hashes_per_tick` before returning the `PohEntry` for the tick.  Subtraction overflow could only happen if two threads where in `tick()` at the same time, which is impossible.

Reply:
hmmmm but multiple consecutive calls to tick() can corrupt the other state that's being tracked like  num_hashes, remaining_hashes, etc. right? So if 

1) Both threads A and B decide they need to tick(), one calls tick(), resets the tracked state. 
2) Thread A calls tick()
3) PohService runs some hashes()
4) Thread B calls tick() and resets the tracked state(num_hashes, remaining_hashes) in poh.rs

Reply:
Well `poh.tick()` must only be called by `poh_recorder.tick()` to ensure that *if* a tick is actually produced that PohRecorder processes in correct order relative to a previous tick() or record().

> Both threads A and B decide they need to tick()

When they decide this, they call `poh_recorder.tick()` and get synchronized.

To clarify the unspoken Poh calling convention ideally the Poh struct would implement two traits I guess: one trait for PohRecorder (with reset,tick,record as methods), and the other trait for PohService (with only Poh.hash()) to advance the inter-tick/record hash.

Reply:
The first line of `poh.hash()` is also subtle and key:
https://github.com/solana-labs/solana/blob/189a6149d6dd740e627721db2fe7ac15bcd85060/core/src/poh.rs#L35

Once PohService reaches one hash before the next tick, continuing to call `poh.hash()` does nothing (`poh.record()` fails at this point in time as well).  To make progress somebody now needs to call `poh.tick()` (via `poh_recorder.tick()`) to unlock the tick PohEntry and get PohService hashing again.  Usually this somebody is PohService but if it's somebody else (like mid-record) then no big deal.

Reply:
oh i see poh_recorder.tick() synchronizes them, makes sense, thanks!

Reply:
hmmm I think I see the intention to use poh_recorder.tick() as the synchronization point, but poh_recorder.tick always calls `poh.tick()` and resets state, for example, setting `self.remaining_hashes = self.hashes_per_tick`.

So for instance in between these two lines here in poh_service: https://github.com/solana-labs/solana/blob/189a6149d6dd740e627721db2fe7ac15bcd85060/core/src/poh_service.rs#L59-L61, 

1) Some banking threads could call `poh_recorder.tick()` and `poh_recorder.record()`
2) Then on line 61, when poh_service calls `tick()`, it will reset the state of the poh, even though poh shouldn't be reset yet.

Reply:
> 2. Then on line 61, when poh_service calls tick(),  it will reset the state of the poh ...

It shouldn't because of 
https://github.com/solana-labs/solana/blob/189a6149d6dd740e627721db2fe7ac15bcd85060/core/src/poh.rs#L69-L71

That is, `poh.tick()` is really `poh.maybe_tick()` -- it'll produce a tick PohEntry only if it's time for a tick.  Otherwise it behaves like `poh.hash(1)`.

Reply:
Oh wow I completely missed that second part of the if statement, my bad haha. Makes total sense, thanks!